irdb/tests/test_publish.py

```diff
@@ -1,104 +1,235 @@
+
+
+@pytest.mark.usefixtures("default_argv")
+@pytest.mark.parametrize("argv, called, args", [
+    ([], False, {}),
+    (["-c"], True, {"stable": False, "keep_version": False}),
+    (["-c", "-s"], True, {"stable": True, "keep_version": False}),
+    (["-c", "-k"], True, {"stable": False, "keep_version": True}),
+    (["-c", "-s", "-k"], True, {"stable": True, "keep_version": True}),
+])
+def test_make_package_called(default_argv, argv, called, args):
+    with mock.patch("sys.argv", default_argv + argv):
+        with mock.patch("irdb.publish.make_package") as mock_mkpkg:
+            pub.main()
+            if called:
+                mock_mkpkg.assert_called_once_with("test_package",
+                                                   args["stable"],
+                                                   args["keep_version"])
+            else:
+                mock_mkpkg.assert_not_called()
+
+
+@pytest.mark.usefixtures("default_argv")
+@pytest.mark.parametrize("argv, called, args", [
+    ([], False, {}),
+    (["-u"], True, {"stable": False, "no_confirm": False}),
+    (["-u", "-s"], True, {"stable": True, "no_confirm": False}),
+    (["-u", "--no-confirm"], True, {"stable": False, "no_confirm": True}),
+    (["-u", "-s", "--no-confirm"], True, {"stable": True, "no_confirm": True}),
+])
+def test_push_to_server_called(default_argv, argv, called, args):
+    pwd = pub.Password("fake_password")
+    with mock.patch("sys.argv", default_argv + argv):
+        with mock.patch("irdb.publish.push_to_server") as mock_phsvr:
+            pub.main()
+            if called:
+                mock_phsvr.assert_called_once_with("test_package",
+                                                   args["stable"],
+                                                   "fake_username", pwd,
+                                                   args["no_confirm"])
+            else:
+                mock_phsvr.assert_not_called()
+
+
+@pytest.mark.usefixtures("default_argv")
+def test_multiple_packages(default_argv):
+    argv = ["foo_package", "bar_package", "-c", "-u"]
+    with mock.patch("sys.argv", default_argv + argv):
+        with mock.patch("irdb.publish.make_package") as mock_mkpkg:
+            with mock.patch("irdb.publish.push_to_server") as mock_phsvr:
+                pub.main()
+                assert mock_mkpkg.call_count == 3
+                assert mock_phsvr.call_count == 3
+                assert "bar_package" in mock_mkpkg.call_args[0]
+                assert "bar_package" in mock_phsvr.call_args[0]
+
+
+@pytest.mark.xfail
+@pytest.mark.usefixtures("default_argv")
+def test_warning_no_action(default_argv):
+    warnmsg = ("Neither `compile` nor `upload` was set. "
+               "No action will be performed.")
+    with mock.patch("sys.argv", default_argv):
+        with mock.patch("sys.stdout", new_callable=StringIO) as mock_stdout:
+            pub.main()
+            assert warnmsg in mock_stdout.getvalue()
```

Comment:
```suggestion
@pytest.mark.usefixtures("default_argv", "capfd")
def test_warning_no_action(default_argv, capfd):
    warnmsg = ("Neither `compile` nor `upload` was set. "
               "No action will be performed.")
    with mock.patch("sys.argv", default_argv):
        pub.main()
        out, err = capfd.readouterr()
        assert warnmsg in out
```
The test also failed on my machine, where I run tests with `python -m pytest`.
The above is based on https://stackoverflow.com/questions/20507601/writing-a-pytest-function-for-checking-the-output-on-console-stdout

Reply:
Perfect, that's the thing I was looking for, thank's a lot! I spent hours over the last few days figuring out how to properly do all these mocks, cause I've never really done that from scratch before 😅 

Still rather new to pytest in general, cases like these are perfect opportunities to learn it 🤓